src/main/java/com/minecolonies/coremod/colony/managers/BuildingManager.java

```diff
@@ -628,20 +628,9 @@ public boolean hasGuardBuildingNear(final IBuilding building)
     @Override
     public void guardBuildingChangedAt(final IBuilding guardBuilding, final int newLevel)
     {
-        for (final IBuilding building : colony.getBuildingManager().getBuildings().values())
+        for (final IBuilding building : getBuildings().values())
         {
-            if (building.getPosition().getX() <= guardBuilding.getPosition().getX() + 16 * guardBuilding.getClaimRadius(newLevel)
-                  && building.getPosition().getZ() <= guardBuilding.getPosition().getZ() + 16 * guardBuilding.getClaimRadius(newLevel))
```

Comment:
If we just fix the distance calculation here we dont need to mark all buildings, just those in the vincinity. It just needs to be a distance calc(2d) of current building compared to the guard building, as the current doesnt work with negative coords

Reply:
As I said, it's not possible to fix the distance calculation without knowing the old level of the building, which is no longer available.  (This is called after the upgrade, so it has access to two copies of the new level and zero copies of the old level. 
 And it can't assume `newLevel - 1` due to deconstruction and creative paste.)

This is because the range it needs to update is the maximum of the old range and the new range.

Reply:
the new range is never lower than the old range

Reply:
It is on deconstruct.

Reply:
For the deconstruct case, the implementation in the PR is more efficient than before, since it defers the calculations until something actually checks the building, instead of recalculating immediately.

Reply:
you can just change the level the guard towers(and barrakcs) pass on deconstruct to this, then we can just mark all in the vincinity for recalc easily